### PR TITLE
sql: Enable ALTER TABLE ... SET LOCALITY REGIONAL BY TABLE

### PR DESCRIPTION
**Pull request by ajstorm**
Release note (sql change): Enables altering of table locality for tables
that are REGIONAL BY TABLE to REGIONAL BY TABLE (but in a different
region).

### CODE REVIEW

**Comment by otan** [comment]
as a workflow thing, can you add me as a reviewer next time?almost missed this cause i don't have a flow for assigned :P

**Review by ajstorm — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59144)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @otan)

___
*[pkg/ccl/logictestccl/testdata/logic_test/alter_table_locality, line 187 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59144#-MRQYbSMH08SMq50DNAn-r1-187:-MRSyFXL5tn7uWYmDNu2:bsg3pwa) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8c4c240d10a8a6854df600e8527e37c265d1cd5/pkg/ccl/logictestccl/testdata/logic_test/alter_table_locality#L187)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

i'd probably get rid of the SHOW ZONE CONFIGURATION here and above, since we already test them elsewhere. but that's optional on you.
</blockquote></details>

I'm going to leave them for now, but will have to take a second pass at this file in a future PR.
___
*[pkg/sql/alter_table_locality.go, line 147 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59144#-MRQYbSMH08SMq50DNAk-r1-147:-MRSxmsT3G6atWT3bbCp:b-zd1npj) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8c4c240d10a8a6854df600e8527e37c265d1cd5/pkg/sql/alter_table_locality.go#L147)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

nit: full stop. some comments above also missing full stops.
</blockquote></details>

I totally need a linter.  I think I've got them all now.
___
*[pkg/sql/alter_table_locality.go, line 241 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59144#-MRQYbSMH08SMq50DNAk-r1-241:-MRSy9qy8vlXmwaepWS3:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8c4c240d10a8a6854df600e8527e37c265d1cd5/pkg/sql/alter_table_locality.go#L241)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

super nit: one lineable:
```
if err := n....; err != nil {
   return err
}
```
</blockquote></details>

done
___
*[pkg/sql/alter_table_locality.go, line 270 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59144#-MRQYbSMH08SMq50DNAk-r1-270:-MRSxOKsCpF4z0XZEiVU:bk08hbx) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8c4c240d10a8a6854df600e8527e37c265d1cd5/pkg/sql/alter_table_locality.go#L270)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

not if we're converting `As`
</blockquote></details>

What's your suggestion here?  Add a TODO to address once AS is in?
___
*[pkg/sql/region_util.go, line 391 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59144#-MRQYbSMH08SMq50DNAl-r1-391:-MRSx6C2AtIulWNSo64H:b-2rkx4d) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8c4c240d10a8a6854df600e8527e37c265d1cd5/pkg/sql/region_util.go#L391)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

i think this is possible with `REGIONAL BY TABLE IN PRIMARY REGION`
</blockquote></details>

Weird.  I'd convinced myself that this was required, and could have sworn I also hit a nil pointer dereference on the Region field.  I did some more digging now and I see that the function being called protects against nil pointers, so I removed this code here.  Good catch!

**Review by otan — APPROVED** [review]
> Reviewed 2 of 4 files at r1, 2 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59144)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajstorm, @arulajmani, and @otan)

___
*[pkg/sql/alter_table_locality.go, line 270 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59144#-MRQYbSMH08SMq50DNAk-r1-270:-MRT-07RDEp_rdvt4ldK:b61mq7g) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8c4c240d10a8a6854df600e8527e37c265d1cd5/pkg/sql/alter_table_locality.go#L270)):*
<details><summary><i>Previously, ajstorm wrote…</i></summary><blockquote>

What's your suggestion here?  Add a TODO to address once AS is in?
</blockquote></details>

return unimplemented in this case, as it is not simply a no-op (yet)

**Review by ajstorm — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59144)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @otan)

___
*[pkg/sql/alter_table_locality.go, line 270 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59144#-MRQYbSMH08SMq50DNAk-r1-270:-MRT0U1t3VbNk6LmLPgD:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8c4c240d10a8a6854df600e8527e37c265d1cd5/pkg/sql/alter_table_locality.go#L270)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

return unimplemented in this case, as it is not simply a no-op (yet)
</blockquote></details>

done

**Comment by ajstorm** [comment]
TFTR!  @arulajmani feel free to have a peek if you want.

bors r=otan

**Comment by otan** [comment]
>	Merge conflict.

Journeys over let's go home

**Comment by ajstorm** [comment]
Nothing like a late night merge.

bors r=otan

**Comment by ajstorm** [comment]
Flakes.  Third time's a charm.

bors r=otan

**Comment by ajstorm** [comment]
Flakes?  Never merge late at night.

bors r=otan

**Comment by ajstorm** [comment]
One more time, then I'm giving up.

bors r=otan

File: pkg/sql/alter_table_locality.go [hunk 1]
```diff
@@ -28,5 +43,246 @@ func (p *planner) AlterTableLocality(
...
+		// by the user. If it has, we need to warn, and only proceed if sql_safe_updates is disabled.
+
+		// Table is placed in the PRIMARY REGION. Remove the table's zone configuration so
+		// that it can be inherited from the database.
+		fullTableName := tree.MakeTableNameWithSchema(
+			tableName.CatalogName,
+			tableName.SchemaName,
+			tableName.ObjectName,
+		)
+		sql := fmt.Sprintf("ALTER TABLE %s CONFIGURE ZONE DISCARD", fullTableName.String())
```
Comment (otan) [inline pkg/sql/alter_table_locality.go:180]: cc @ajstorm late night epiphanies just were had

just realised we probably don't want to discard here if the user modified other fields (e.g. range_min_bytes). can you add a TODO in a subsequent PR?

**Review by ajstorm — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59144)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @otan)

___
*[pkg/sql/alter_table_locality.go, line 180 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/59144#-MRQYbSMH08SMq50DNAk-r3-180:-MR_44_uBhU9tvzjTAXz:bz9w5ax) ([raw file](https://github.com/cockroachdb/cockroach/blob/0fdca98e96471884fd91e86da6fa456fa091a1f8/pkg/sql/alter_table_locality.go#L180)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

cc @ajstorm late night epiphanies just were had

just realised we probably don't want to discard here if the user modified other fields (e.g. range_min_bytes). can you add a TODO in a subsequent PR?
</blockquote></details>

Ahh, Reviewable...

You mean this TODO that was in the "Collapsed" part?  

		// TODO(#multiregion): We should check to see if the zone configuration has been updated
		// by the user. If it has, we need to warn, and only proceed if sql_safe_updates is disabled.

**Comment by otan** [comment]
___
*[pkg/sql/alter_table_locality.go, line 180 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/59144#-MRQYbSMH08SMq50DNAk-r3-180:-MRaJ5gZ8JFVpdueFYD9:b-y93bbm) ([raw file](https://github.com/cockroachdb/cockroach/blob/0fdca98e96471884fd91e86da6fa456fa091a1f8/pkg/sql/alter_table_locality.go#L180)):*
<details><summary><i>Previously, ajstorm wrote…</i></summary><blockquote>

Ahh, Reviewable...

You mean this TODO that was in the "Collapsed" part?  

		// TODO(#multiregion): We should check to see if the zone configuration has been updated
		// by the user. If it has, we need to warn, and only proceed if sql_safe_updates is disabled.

</blockquote></details>

hah! yeah that's good nuff for me

**Pull request merged**